Add API end-points for review/approval workflow

punchagan opened the pull request.

tangrammer commented on backend/resources/migrations/080-add-topic-review-table.up.sql:
```diff
@@ -0,0 +1,27 @@
+  topic_name topic_name NOT NULL,
+  topic_id integer NOT NULL,
+  assigned_by integer NOT NULL REFERENCES stakeholder(id),
+  assigned timestamptz DEFAULT now(),
```
it seems strange not use `created` as column name instead of `assigned`... also having all the tables same column name to track db-entity creation-date

punchagan replied:
Fair!

tangrammer commented on backend/resources/migrations/080-add-topic-review-table.up.sql:
```diff
@@ -0,0 +1,27 @@
+  assigned_by integer NOT NULL REFERENCES stakeholder(id),
+  assigned timestamptz DEFAULT now(),
+  reviewer integer NOT NULL REFERENCES stakeholder(id),
+  reviewed timestamptz DEFAULT NULL,
```
same here ... it seems like we are defining here the `modified` column ... unless we want to track time with other meaning

punchagan replied:
I was trying to capture the time when a review was submitted, and not necessarily when this row was modified last. But, I think we could switch to using a `modified` field for consistency.

tangrammer commented on backend/src/gpml/constants.clj:
```diff
@@ -7,4 +7,6 @@
    (sort
     (apply conj resource-types ["event" "technology" "policy" "project" "stakeholder" "organisation"]))))
 
+(def reviewer-review-status ["PENDING" "ACCEPTED" "REJECTED"])
```
perhaps we can start to use the clojure way for enums  https://stackoverflow.com/a/11856228

tangrammer commented on backend/src/gpml/db/review.sql:
```diff
@@ -0,0 +1,31 @@
+UPDATE review SET
+  review_status = :review-status::reviewer_review_status,
+  review_comment = :review-comment
+  WHERE id = :id RETURNING id;
```
should we add modified/reviewed column time value?

punchagan replied:
I'll use the DB trigger we have for this, instead.

tangrammer replied:
where is the trigger? in this PR?

punchagan replied:
https://github.com/akvo/unep-gpml/pull/386/commits/4210138e8df0831f9591cab1d1fbb34351c2a9e6#diff-6c9231224a44aeacfe90dff06e9a6434c8dfaf82bea51bebd905893bce2d97a2R28

tangrammer commented on backend/src/gpml/auth.clj:
```diff
@@ -117,3 +117,11 @@
+  (fn [handler]
+    (fn [{:keys [user approved?] :as request}]
+      (if (and approved? (contains? #{"ADMIN" "REVIEWER"} (:role user)))
+        (handler (assoc request :admin user))
```
IMO it would be more specific if we use `:reviewer` 
`(assoc request :reviewer user)`

tangrammer commented on backend/resources/gpml/config.edn:
```diff
@@ -84,13 +84,45 @@
-                                                                 #ig/ref :gpml.auth/approved-user
-                                                                 #ig/ref :gpml.auth/admin-required-middleware]}
+                                                                 #ig/ref :gpml.auth/approved-user]}
                                          ["/get-reviewers"
```
as we agree in zulip, move this entrypoint to "/reviewer"

punchagan replied:
Done.

tangrammer commented on backend/resources/gpml/config.edn:
```diff
@@ -84,13 +84,45 @@
+                                                                       [:review-comment string?]
+                                                                       [:review-status #ig/ref :gpml.handler.review/review-status]]}
+                                                   :handler #ig/ref :gpml.handler.review/update-review}}]
+                                         ["/assign-reviewer/{topic-type}/{topic-id}"
```
Thinking in taking out from `/submission` and use also `/reviewer`
`POST /reviewer/topic-type/topic-id` 
BTW: why here (API) the first arg is called `topic-type` but in backend and db is called `topic-name` ... should we unify names here?

punchagan replied:
Agree on unifying names for `topic-type` and `topic-name`.  

The `assign-reviewer` is currently in `/review` and not `/submission`. So, I'm not sure what your entire suggestion here is. :)

tangrammer commented on backend/resources/gpml/config.edn:
```diff
@@ -243,6 +275,10 @@
 
   :gpml.handler.review/get-reviewers {:db #ig/ref :duct.database/sql}
+  :gpml.handler.review/get-review {:db #ig/ref :duct.database/sql}
+  :gpml.handler.review/assign-reviewer{:db #ig/ref :duct.database/sql}
```
perhaps new-reviewer/create-reviewer instead of assign-reviewer?

tangrammer commented on backend/src/gpml/handler/review.clj:
```diff
@@ -1,13 +1,86 @@
+
+(defmethod ig/init-key ::get-review [_ {:keys [db]}]
+  (fn [{{{:keys [topic-type topic-id]} :path} :parameters}]
+    (get-review db topic-type topic-id)))
```
why we don't include de implementation inside these defmethods instead of extracting the logic to a new function?

punchagan replied:
It seems easier to (re) evaluate the function, when developing and modifying them. We'd otherwise have to `reset` the system, when we make changes to these functions. We could move the code into the `defmethod`s all across the codebase, if we think that's a better way to define them.

tangrammer commented on backend/src/gpml/handler/review.clj:
```diff
@@ -1,13 +1,86 @@
+(defn get-review [db topic-type topic-id]
+  (let [conn (:spec db)
+        topic-name (cond
+                     (contains? (set constants/resource-types) topic-type) "resource"
```
perhaps it's time to use `set` in `gpml.constants` ... I found in the code lots of `(set constants/resource-types)`

tangrammer commented on backend/src/gpml/handler/review.clj:
```diff
@@ -1,13 +1,86 @@
+                     (contains? (set constants/resource-types) topic-type) "resource"
+                     (= topic-type "project") "initiative"
+                     :else topic-type)]
+    (jdbc/with-db-transaction [conn (:spec db)]
```
a refactor proposal 
```clojure
(defn assign-reviewer [db topic-type topic-id reviewer assigned-by]
  (let [topic-name (cond
                     (contains? (set constants/resource-types) topic-type) "resource"
                     (= topic-type "project") "initiative"
                     :else topic-type)]
    (jdbc/with-db-transaction [conn (:spec db)]
      (resp/response (if-let [review (db.review/review-by-topic-item
                                      conn
                                      {:topic-name topic-name :topic-id topic-id})]
                       (db.review/change-reviewer conn {:id          (:id review)
                                                        :assigned-by assigned-by
                                                        :reviewer    reviewer})
                       (db.review/new-review conn {:topic-name  topic-name
                                                   :topic-id    topic-id
                                                   :assigned-by assigned-by
                                                   :reviewer    reviewer}))))))
```

tangrammer commented on backend/src/gpml/handler/review.clj:
```diff
@@ -1,13 +1,86 @@
+        (resp/response updated)))))
+
+(defn update-review [db topic-type topic-id review-status review-comment email]
+  (let [topic-name (cond
```
this logic is called twice in same file ... we should extract it to its own function

punchagan replied:
Agree!

tangrammer commented on backend/src/gpml/handler/review.clj:
```diff
@@ -1,13 +1,86 @@
+                     (= topic-type "project") "initiative"
+                     :else topic-type)
+        current-user (db.stakeholder/stakeholder-by-email (:spec db) {:email email})]
+    (jdbc/with-db-transaction [conn (:spec db)]
```
refactor proposal
```clojure
(defn update-review [db topic-type topic-id review-status review-comment email]
  (let [topic-name (cond
                     (contains? (set constants/resource-types) topic-type) "resource"
                     (= topic-type "project") "initiative"
                     :else topic-type)
        current-user (db.stakeholder/stakeholder-by-email (:spec db) {:email email})
        res403 {:status 403 :body {:message "Cannot update review for this topic"}}]
    (jdbc/with-db-transaction [conn (:spec db)]
      (if-let [review (db.review/review-by-topic-item
                       conn
                       {:topic-name topic-name :topic-id topic-id})]
        (if (not (= (:reviewer review) (:id current-user)))
          (resp/response (db.review/update-review
                          conn
                          {:id (:id review)
                           :review-status review-status
                           :review-comment review-comment}))
          res403)
        res403))))
```

tangrammer reviewed (commented):
Hi @punchagan 
I left a few (perhaps too many 😬 ) comments on this PR, perhaps you prefer me to add these changes to the code instead of doing you the required work 
let me know if you are busy with other stuff and I can help on this PR more

punchagan commented:
Thanks for your detailed review @tangrammer! I addressed most of the comments except one about changing "assign-reviewer" to something like "new-review", and moving it to the `/reviewer` namespace. Could you elaborate a little more on that?

tangrammer commented on backend/src/gpml/constants.clj:
```diff
@@ -1,10 +1,12 @@
 (ns gpml.constants)
 
-(def resource-types ["financing_resource" "technical_resource" "action_plan"])
+(def resource-types #{"financing_resource" "technical_resource" "action_plan"})
```
just a doble check here... are we returning this data directly to the FE? in that case do we need order there? in that case 😄 perhaps we need 2 defs, `(def resource-types ["financing_resource" "technical_resource" "action_plan"])` and `(def resource-types-set (set resource-types))`

punchagan replied:
I don't think we are returning it to the FE directly. I did consider adding a new constant, but it seems like the main use is to check containership, and one use where we do the same thing in a DB query.

tangrammer reviewed (commented):
Hi Puneeth, I left again a couple of tiny comments, I think we are almost done 😄

tangrammer reviewed (approved):
I'm pretty happy with the changes!

punchagan commented:
Thanks again for the detailed review and great suggestions/refactors for improvement.
